java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/utils/TypedValueBinder.java

```diff
@@ -0,0 +1,126 @@
+  }
+
+  /**
+   * Bind the given Avatica values to the prepared statement.
+   * @param typedValues The parameter values.
+   */
+  public void bind(List<TypedValue> typedValues) {
+    if (preparedStatement.getParameterSchema().getFields().size() != typedValues.size()) {
+      throw new IllegalStateException(
+          String.format("Prepared statement has %s parameters, but only received %s",
+              preparedStatement.getParameterSchema().getFields().size(),
+              typedValues.size()));
+    }
+
+    for (int i = 0; i < typedValues.size(); i++) {
+      final TypedValue param = typedValues.get(i);
+      final FieldVector vector = parameters.getVector(i);
+      bindValue(param, vector);
+    }
+
+    if (!typedValues.isEmpty()) {
+      parameters.setRowCount(1);
+      preparedStatement.setParameters(parameters);
+    }
+  }
+
+  private void bindValue(TypedValue param, FieldVector vector) {
+    Object value = param.value;
+    try {
+      if (vector instanceof IntVector) {
+        ((IntVector) vector).setSafe(0, (int) value);
+      } else if (vector instanceof BigIntVector) {
+        BigIntVector longVec = (BigIntVector) vector;
+        if (value instanceof Long) {
+          longVec.setSafe(0, (long) value);
+        } else {
+          longVec.setSafe(0, (int) value);
+        }
+      } else if (vector instanceof BitVector) {
+        ((BitVector) vector).setSafe(0, (int) value);
+      } else if (vector instanceof Float4Vector) {
+        ((Float4Vector) vector).setSafe(0, (float) value);
+      } else if (vector instanceof Float8Vector) {
+        ((Float8Vector) vector).setSafe(0, (double) value);
+      } else if (vector instanceof DecimalVector) {
+        ((DecimalVector) vector).setSafe(0, (BigDecimal) value);
+      } else if (vector instanceof VarCharVector) {
+        ((VarCharVector) vector).setSafe(0, new Text((String) value));
+      } else if (vector instanceof LargeVarCharVector) {
+        ((LargeVarCharVector) vector).setSafe(0, new Text((String) value));
+      } else if (vector instanceof DateMilliVector) {
+        String pattern = "yyyy-MM-dd HH:mm:ss.SSS";
+        DateTime dateTime = DateTime.parse((String) value, DateTimeFormat.forPattern(pattern));
+
+        ((DateMilliVector) vector).setSafe(0, dateTime.getMillis());
+      } else if (vector instanceof VarBinaryVector) {
+        ((VarBinaryVector) vector).setSafe(0, (byte[]) value);
+      } else if (vector instanceof LargeVarBinaryVector) {
+        ((LargeVarBinaryVector) vector).setSafe(0, (byte[]) value);
+      } else if (vector instanceof NullVector) {
+        assert true;
+      } else {
+        throw new UnsupportedOperationException(
+                String.format("Binding to parameter of Arrow type %s", vector.getField().getType()));
+      }
+    } catch (ClassCastException e) {
+      throw new RuntimeException(String.format("Value of type %s is not compatible with Arrow type %s",
```

Comment:
This isn't truly an error. In JDBC, a user is allowed to bind to a different type that the server expects and the driver is supposed to do the conversion behind-the-scenes. This doesn't happen that often. Applications sometimes do this with integers where they use a 32-bit integer regardless of what type the server is expecting.

Reply:
ahhh interesting.. that makes sense. Should we somehow keep track of the `Fields` list that the server returns then to make sure we're casting to the expected type?

We have this strange dance here: of Server

Reply:
Yes, it'd be good to keep track of the expected arrow type.

Reply:
Hmm lol ok nvm, the `Field` list we have here already IS the one returned from the server.

> a user is allowed to bind to a different type that the server expects and the driver is supposed to do the conversion behind-the-scenes

Interesting... so does that mean we need to be a whole lot smarter about casting the `TypedValue` to a Java type? Doesn't Avatica already take care of this for us?

I added that exception handler in case the casting when setting the value failed at least there would be a useful error, but I wouldn't expect to hit that case since the AvaticaParameter should already specify and handle the expected type.

Reply:
Avatica might handle this -- haven't tried it.

Reply:
After digging a bit, it looks like Avatica does not really do any verification that the obect that is set by the user is compatible with type defined for the `AvaticaParameter`. This likely means we'll need more complex type conversion/validation so I refactored to use a `ArrowType.ArrowTypeVisitor` pattern for both generating the `AvaticaParameter` and for binding the parameter value to a Vector.

This helps us accomplish a couple of things:
- Ensure all Arrow types are covered
- Keep all conversion logic between AvaticaParameters/TypedValues and Arrow in the same class (one class per ArrowType)

For the purpose of this PR, I think we should just assume the user is setting the correct type and simply throw an error if we can't cast correctly. This conversion logic can easily be improved over time as needed.

Reply:
This is fine for me. I'd suggest we raise an issue to add mismatched user parameter -> Flight server type conversion

Reply:
Yeah right now I have this:

```
    } catch (ClassCastException e) {
      throw new RuntimeException(String.format("Value of type %s is not compatible with Arrow type %s",
              typedValue.type, vector.getField().getType()));
    }
```

Happy to change the exception type or message.

Reply:
I reworded this a little. See `AvaticaParameterBinder.bind(vector, typedVale, index)`